### PR TITLE
Mirror rowan architecture in our CST

### PR DESCRIPTION
**Pull request by AntonyBlakey**

### CODE REVIEW

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cst.rs:14]: suggestion: Maybe "offset" instead of "size", as it is almost always used by users as offsets (and in ranges).

Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/cst.rs:14]: we can then hide/not expose `text_size` at all, and always produce offsets, ranges, starts and ends to the user through our APIs.
`TextSize` type can be `TextOffset` as well, to work with `TextRange<>`

File: crates/codegen/syntax_templates/src/rust/cst_visitor.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/rust/cst_visitor.rs:11]: I think users are expecting to see ranges, and they will need it to store/pass to editors/linters.
What do you think about providing that instead?

```suggestion
        range: TextRange,
```

File: .changeset/shy-carrots-love.md
Comment (OmarTawfik) [inline .changeset/shy-carrots-love.md:5]: I suggest wording changesets in terms of user-visible changes, rather than for example "refactoring parser", as it won't change the public API/behavior of the parser. For example, mentioning that CST nodes now have a different structure, or not storing ranges on nodes for perf, or tokens carrying the indvididual strings now, etc...
We can also include more than one changeset (markdown file) in this PR, and they will be aggregated as multiple entries in the `CHANGELOG.md`.

File: crates/codegen/syntax_templates/src/rust/cst_visitor.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/rust/cst_visitor.rs]: done

File: .changeset/shy-carrots-love.md
Comment (AntonyBlakey) [reply .changeset/shy-carrots-love.md]: dpne

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cst.rs:13]: Nit suggestion: What do you think of `length` instead of `text_len`? The type already starts with "Text". Or the full `text_length` if not.

```suggestion
    pub length: TextIndex,
```

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/cst.rs:57]: Is this intended to be part of the public API? I wonder what does it offer over `node.text_len().utf8`?

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/parse_error.rs:9]: Not sure I understand why it is diferent than `expected_tokens`. Is it used differently?

```suggestion
        expected_tokens: Vec<TokenKind>,
```

File: crates/codegen/syntax_templates/src/typescript/cst_types.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/typescript/cst_types.rs:100]: Is this meant to be exposed to te public API? does it need `#[napi]`?
If so, I suggest adding a test to get the object back and assert its JSON.

File: crates/codegen/syntax_templates/src/shared/parser_result.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/parser_result.rs:77]: I love this!

File: crates/codegen/syntax_templates/src/rust/parse_output.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/rust/parse_output.rs:27]: I wonder why are we not using a `TextIndex` here. Since we expose this information to the public API, shouldn't we expose the other indices as well?

File: crates/codegen/syntax_templates/src/typescript/parse_output.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/typescript/parse_output.rs:43]: Same question here, about using `Range<TextIndex` and exposing the other types.

File: crates/codegen/syntax_templates/src/shared/text_index.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/text_index.rs:15]: do we intend to expose similar utilities for typescript as well? If not, I wonder if we should make it private until we think about it.

File: crates/codegen/syntax_templates/src/shared/parser_helpers.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/parser_helpers.rs:7]: At the risk of being repetitive, I absolutely love this! :D

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/package/src/generated/index.d.ts:652]: I think the return type regressed here? Is this intentional?

File: crates/solidity/testing/utils/src/cst_snapshots/mod.rs
Comment (OmarTawfik) [inline crates/solidity/testing/utils/src/cst_snapshots/mod.rs:125]: super nit: these two are perfect candidates for the "range_utf8" util added earlier.

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs
Comment (OmarTawfik) [inline crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:14]: should we remove this?

File: .changeset/shy-carrots-love.md
Comment (OmarTawfik) [inline .changeset/shy-carrots-love.md:5]: Looks like CI is failing because of misspelled `containe` here.

File: Cargo.lock
Comment (OmarTawfik) [inline Cargo.lock:682]: Not sure why the version was changed here, while we are still using the old version in `Cargo.toml`. Is this intentional?

Cargo should have errored out here. I think I'm missing something.

File: crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml:11]: Nothing is skipped in this case. Is the zero-length token intentional?

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml:8]: I think this is a regression in the error messages?

File: crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml:11]: I noticed that when we create an error, we specify a range from current position till end of document. That can be quite large if the error is in the beginning of the file.

I realize we need to move to a DFA tokenizer for us to use the range of the current token. But until we get there, I wonder if it is better to limit it somehow, for example, simply clipping it to the end of the current line.

Otherwise, something like the HardHat LSP would paint the entire user window/file with red squiggles, making it much harder to read the code.

File: crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/generated/0.5.0.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/generated/0.5.0.yml:8]: this exposed a bug in our grammar where `0` followed by `x` should not be matchd as a `DecimalLiteral`, but as a `HexLiteral` .. I will fix it in my current PR.

But I wonder what changed the output in this PR though? Are we scanning input differently somehow?

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml:9]: Is this error location correct? It is expecting `FallbackKeyword or Identifier or ReceiveKeyword` at the beginning of the file, instead of at the end.

File: crates/solidity/testing/snapshots/cst_output/HexStringLiteral/invalid_leading_separator/generated/0.4.11.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/HexStringLiteral/invalid_leading_separator/generated/0.4.11.yml:7]: I wonder if that also fixes #335?

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (OmarTawfik) [inline crates/codegen/syntax_templates/src/shared/parse_error.rs:15]: suggestion for these places and others, using the shorter syntax for compactness:

```suggestion
            utf8_range: utf8_position..utf8_position,
```

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml:18]: Earlier, we discovered a bug where operators don't wrap their operands with an additional base `Expression` node. This is correct if `Expression` is inlined, but in this case here, it is not, and therefore, they should be wrapped.
Otherwise, users won't be able to visit all expressions recusively.

IIRC, it was getting fixed in this PR, is that still the plan?
Definetely not blocking if not.

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Fantastic work!

I think there is a couple of test regressions that we might want to address before merging. And also some new APIs that I'm not sure if it was intentional to expose them, or if we should keep them private for now. Otherwise, looks great.

File: .changeset/shy-carrots-love.md
Comment (AntonyBlakey) [reply .changeset/shy-carrots-love.md:5]: done

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml:18]: Currently, operators don't wrap their operands in an `Expression` node.
For inlined nodes, this should be correct, but for non-inlined ones (like `Expression` here), we should be wrapping them. Otherwise, visitors won't be able to iterate on all expressions recursively.

Since this affects CST stabilization, I wonder if it is a quick/easy fix to do in this PR, or would we have to postpone for afterwards?

File: Cargo.lock
Comment (AntonyBlakey) [reply Cargo.lock:682]: no idea, not my doing.

File: crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/generated/0.5.0.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/generated/0.5.0.yml:8]: This is an uppercase `X`, which from 0.5.0 onwards is not a valid hex literal

File: crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs
Comment (AntonyBlakey) [reply crates/solidity/testing/utils/src/cst_snapshots/test_nodes.rs:14]: done

File: crates/codegen/syntax_templates/src/rust/parse_output.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/rust/parse_output.rs:27]: done

File: crates/codegen/syntax_templates/src/typescript/parse_output.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/typescript/parse_output.rs:43]: done

File: crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml:11]: Yes, because I need some token to be returned. Not happy with it aesthetically. I think it should maybe be an EOF token. Refactor it later.

File: crates/codegen/syntax_templates/src/typescript/cst_types.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/typescript/cst_types.rs:100]: no, it's just an internal conversion routine to wrap nodes into their TS objects

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cst.rs:13]: I think `text_len` is the best choice because:
1. `len` is the rust standard for length methods
2. Just `len` or `length` would conceivably be the number of childred nodes
3. Which is probably why rowan does it.

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/parse_error.rs:9]: Changed

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/parse_error.rs:15]: done

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts
Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/package/src/generated/index.d.ts:652]: No, it wasn't :)

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml:8]: This is a known issue with the conditions at the boundary of the pratt and LL parsers. No solution yet.

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml:18]: I forgot we discussed this. I would rather fix this in a small PR immediately after this one.

File: crates/solidity/testing/snapshots/cst_output/HexStringLiteral/invalid_leading_separator/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/HexStringLiteral/invalid_leading_separator/generated/0.4.11.yml:7]: Yes, because we now fail tokens in their entirity

File: crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml:11]: I think we should do the simple (correct) thing, and leave that to the LSP author to maybe elide the erro range.

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml:9]: That's because the error range covers the entire production IIUC.

File: crates/solidity/testing/utils/src/cst_snapshots/mod.rs
Comment (AntonyBlakey) [reply crates/solidity/testing/utils/src/cst_snapshots/mod.rs:125]: done

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/parse_error.rs:11]: I still see 146 instances of `tokens_that_would_have_allowed_more_progress`. I think we missed  a few.

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/cst.rs:57]: Should we remove `utf8_len` now we have `TextRangeExtensions`?

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/parse_error.rs:11]: Actually, `expected_tokens` should be changed, I missed this. `expected_tokens` is IMO not sufficiently expressive. Having a meaningful name really made this easier to think about

File: crates/codegen/syntax_templates/src/shared/cst.rs
Comment (AntonyBlakey) [reply crates/codegen/syntax_templates/src/shared/cst.rs:57]: Will do

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts
Comment (OmarTawfik) [reply crates/solidity/outputs/npm/package/src/generated/index.d.ts]: For the `[utf8: number, utf16: number, char: number]` type, I wonder at what time is it better to use an object `TextIndex` with named properties (`utf8`/`utf16`/`char`) over the tuple. Is it cheap to serialize the object to TS side at once, given that we don't need to use identity or send it back to rust.

Not blocking for this PR though.

Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/package/src/generated/index.d.ts]: I had the same thought, particularly after adjust the TS test and seeing that I had to use array indices.

File: crates/codegen/syntax_templates/src/shared/parse_error.rs
Comment (OmarTawfik) [reply crates/codegen/syntax_templates/src/shared/parse_error.rs:11]: I'm fine with using the other name everywhere instead. Just the two names confused me, as I was tracing when does it carry different values.

File: Cargo.lock
Comment (OmarTawfik) [reply Cargo.lock]: I see it was removed from this PR. Should be fine now.

File: crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/ContractDefinition/zero_length_input/generated/0.4.11.yml:11]: #507

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml:8]: I suggest bringing back the `end of input` string to `render_error_report()` in `parse_error.rs` if `expectations` is empty (even if temporarily)? Otherwise the error string is confusing.

File: crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/MappingType/named_both/generated/0.4.11.yml:11]: Sounds good.

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/Expression/function_call_options/generated/0.4.11.yml:8]: Good idea, will do

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml:9]: The error is produced while parsing the `FunctionDefinition` production, not `SourceUnit`. Am I missing something?

File: crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/Expression/function_call_options_split/generated/0.8.0.yml:18]: #508

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml
Comment (AntonyBlakey) [reply crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml:9]: I will have to check, but that shows a across-multiple-lines error, so I'll have to investigate. I'll create an issue for this because IMO not worth holding up this PR.

File: crates/solidity/outputs/npm/package/src/generated/index.d.ts
Comment (OmarTawfik) [reply crates/solidity/outputs/npm/package/src/generated/index.d.ts]: #509

File: crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/SourceUnit/partial_definition/generated/0.4.11.yml:9]: #510

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**
